docs/dev/testing.md

```diff
@@ -90,32 +99,162 @@ reject file:
 must thorougly examine the diff and understand the reasons for every
 change before overwriting `.result` files.
 
+### Debugging CQL tests
+
+To debug CQL tests, one can run CQL against a standalone
+scylla (possibly started in debugger) using `cqlsh`.
+
 ## Unit tests
 
 The same unit test can be run in different seastar configurations, i.e. with
 different command line arguments. The custom arguments can be set in
 `custom_args` key of the `suite.yaml` file.
 
-
-## Debugging
+### Debugging unit tests
 
 If a test fails, its log can be found in `testlog/${mode}/testname.log`.
 By default, all unit tests are built stripped. To build non-stripped tests,
 `./configure` with `--tests-debuginfo list-of-tests`.
 `test.py` adds some command line arguments to unit tests. The exact way in
 which the test is invoked is recorded in `testlog/test.py.log`.
 
-To debug CQL tests, one can invoke `gdb cql_repl`, which will start
-`cql_repl` in interactive mode, and then supply faulty CQL statements
-on the terminal.
-
-
-## Automation
+## Python tests
+
+`test.py` supports pytest standard of tests, for suites (directories)
+specifying 'Python' test type in their scylla.yaml. For such tests,
+a standalone server instance is created, and a connection URI to the
+server is passed to the test. Thanks to conveninence fixtures,
+test writers don't need to create or cleanup connections or keyspaces.
+`test.py` will also keep track of the used server(s) and will shut
+down the server when all tests  using it end.
+
+## Sharing and pooling servers
+
+Since there can be many pytests in a single directory (e.g. cql-pytest)
+`test.py` creates multiple servers to parallelize their execution.
+Each server is also shared among many tests, to save on setup/teardown
+steps. While this speeds up execution, sharing servers complicates debugging
+if a test fails.
+
+Specifically, you should avoid leaving global artifacts in your test, even
+if it fails. Typically, you could use a built-in `keyspace()` fixture
+to create a randomly named keyspace.
+
+At start and end of each test, `test.py` performs a number of sanity checks
+of the used server:
+- it should be up and running,
+- it should not contain non-system keyspaces.
```

Comment:
I never noticed we had this check. Do we?
I'm a bit worried about it. It means that if one of our tests has a bug that it doesn't clean up after itself, the immediate penalty will be starting a new Scylla server and slowing down the test. The author of the messy test will not even know of this penality, so what's the point? Isn't the fact that tests use a random keyspaces, random tables and when tables are shared - random keys - more than enough, and it's actually not a problem if one test left something behind?

Reply:
We count the total number of keyspaces at start, and at end checks it remains the same.

Reply:
Ok. I think this test may be a mistake (for the reasons I outlined below), but of course we can't change it in the documentation patch, and should document the code we have.